server/common/fbs/fbs_coders.py

```diff
@@ -0,0 +1,225 @@
+import server.common.fbs.NetEncoding.SparseFloat32FBArray as SparseFloat32FBArray
+import server.common.fbs.NetEncoding.SparseFloat64FBArray as SparseFloat64FBArray
+
+
+def _get_array_class(array):
+    # return the generic type of array - category, sparse, or dense
+    # this determines which coder should be used.
+    if isinstance(array, pd.Series) and array.dtype.name == "category":
+        return "category"
+    elif sp.issparse(array):
+        return "sparse"
+    else:
+        return "dense"
+
+
+def serialize_typed_array(builder, source_array):
+    if isinstance(source_array, pd.Index):
+        source_array = source_array.to_series()
+
+    array_class = _get_array_class(source_array)
+
+    # depending on the array class, compute the encoding data type from
+    # the corresponding attribute of the source array. e.g. for pandas Categoricals,
+    # the encoding data type is computed from the codes array.
+    if array_class == "category":
+        encoding_dtype = np.dtype(source_array.cat.codes.values.dtype).str
+    elif array_class == "sparse":
+        encoding_dtype = np.dtype(get_encoding_dtype_of_array(source_array.data)).str
+    else:
+        encoding_dtype = np.dtype(get_encoding_dtype_of_array(source_array)).str
+
+    # two-layer mapper (1) array_class, (2) encoding_dtype
+    # this is necessary as an int32 codes array will require a different coder
+    # than an int32 numeric array.
+    arrayEncoder = {
+        "dense": {
+            np.dtype(np.float64).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Float32FBArray),
+            np.dtype(np.float32).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Float32FBArray),
+            np.dtype(np.float16).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Float32FBArray),
+            np.dtype(np.int8).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Int32FBArray),
+            np.dtype(np.int16).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Int32FBArray),
+            np.dtype(np.int32).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Int32FBArray),
+            np.dtype(np.int64).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Int32FBArray),
+            np.dtype(np.uint8).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Uint32FBArray),
+            np.dtype(np.uint16).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Uint32FBArray),
+            np.dtype(np.uint32).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Uint32FBArray),
+            np.dtype(np.uint64).str: (DenseNumericCoder, TypedFBArray.TypedFBArray.Uint32FBArray),
+        },
+        "sparse": {
+            np.dtype(np.float64).str: (SparseNumericCoder, TypedFBArray.TypedFBArray.SparseFloat64FBArray),
+            np.dtype(np.float32).str: (SparseNumericCoder, TypedFBArray.TypedFBArray.SparseFloat32FBArray),
+            # below is to handle cases where integer sparse arrays are encoded - densify in this case.
```

Comment:
> because the matrix FBS encoding doesn't support sparse int arrays, because we have not yet found a use for them? 

Both. Expressions are the only thing we are sparse coding and they're floats. I'll clarify with a comment! Alternatively, I could add support for sparse int arrays, which may be useful if we ever wish to accommodate integer-typed raw counts, but that could also happen in a future PR. I would go with a clarifying comment for now.